src/map/utils/battleutils.cpp

```diff
@@ -2210,8 +2210,16 @@ namespace battleutils
 
             // add tp to attacker
             if (primary)
-            {
-                PChar->addTP((int16)(((tpMultiplier * baseTp) + bonusTP) * (1.0f + 0.01f * (float)((PChar->getMod(Mod::STORETP) + getStoreTPbonusFromMerit(PChar))))));
+
+            // Check if SAVETP value is more than WS would return and use it instead if so.
+            { 
+                int16 standbyTp = ((int16)(((tpMultiplier * baseTp) + bonusTP) * (1.0f + 0.01f * (float)((PChar->getMod(Mod::STORETP) + getStoreTPbonusFromMerit(PChar))))));
+                int16 savedTp = (PChar->getMod(Mod::SAVETP));
+
+                if (savedTp >= standbyTp) 
```

Comment:
this part (the block that adds the TP) needs to be outside the `damage > 0` block (line 2240 ish) or else it won't apply on misses.  additionally, you can shorten the adding to simply `PChar->addTP(std::max(savedTp, standbyTp));`

Reply:
Thanks for the feedback. Will test suggestion and submit an additional commit after testing code.